CIPs/CIP-122.md

```diff
@@ -0,0 +1,454 @@
+
+
+## Motivation
+
+<!--Motivation is critical for CIPs that want to change the Ceramic protocol. It should clearly explain why the existing protocol specification is inadequate to address the problem that the CIP solves. CIP submissions without sufficient motivation may be rejected outright.-->
+Currently in Ceramic the main accounts types are PKH DID. These are great because they enable existing wallets to be used directly with Ceramic. However, these accounts are inherently tied to a specific account and there is no way to rotate keys in a simple manner. The 3 DID method changes this by defining a DID method which works with object-capabilities and supports all DID CRUD operations.
+
+Furthermore, the revocation regsitry of the 3 DID method enables the DID subject to revoke any object-capability issued by the 3ID. This can be useful in case of a key compromise, e.g. for a temporary session key.
+
+## Specification
+
+<!--The technical specification should describe the syntax and semantics of any new feature.-->
+Specification goes here.
+
+### The did:3 identifier
+
+> ```
+> did:3:<cacao-mh>
+> ```
+
+The `<cacao-mh>` value is a `keccak-256` multihash over the CID of a CACAO object-capability, e.g. ReCap or UCAN. The value MUST be encoded as a multibase string, using `base58btc`.
+
+***Simple example:***
+
+> ```
+> did:3:z9CEZ1N1gbFK8J3rxVw3o6M5wygjoNFRSaEtkoGZw5fmbj
+> ```
+
+### CRUD Operation Definitions
+
+3IDs are created, updated, and deactivated by creating, notarizing, and revoking object-capabilities in the form of CACAOs.
+
+#### Create
+
+Create and sign a [SIWx](https://chainagnostic.org/CAIPs/caip-122) message where:
+
+- `address` - a blockchain address (e.g. `0xc02aaa39b223fe8d0a0e5c4f27ead9083c756cc2`)
+- `uri` - the PKH DID for the same address (e.g. `did:pkh:eip155:1:0xc02aaa39b223fe8d0a0e5c4f27ead9083c756cc2`)
+
+And resources contains the following ReCap object:
+
+```java
+{
+  "att": {
+    "did:3:new": { "3id/control": [] }
+  }
+}
+```
+
+- *When signing this would look something like this for the user:*
+
+  ```
+  example.com wants you to sign in with your Ethereum account:
+  0xc02aaa39b223fe8d0a0e5c4f27ead9083c756cc2
+  
+  I further authorize https://example.com to perform the following 
+  actions on my behalf: (1) "3id/control" for "did:3:new".
+  
+  URI: did:pkh:eip155:1:0xc02aaa39b223fe8d0a0e5c4f27ead9083c756cc2
+  Version: 1
+  Chain ID: 1
+  Nonce: n-0S6_WzA2Mj
+  Issued At: 2022-06-21T12:00:00.000Z
+  Resources:
+  - urn:recap:example:eyJkZWYiOlsicmVhZCJdLCJ0YXIiOnsibXkucmVzb3VyY2UuMSI6WyJhcHBlbmQiLCJkZWxldGUiXSwibXkucmVzb3VyY2UuMiI6WyJhcHBlbmQiXSwibXkucmVzb3VyY2UuMyI6WyJhcHBlbmQiXX19
+  ```
+
+##### Create the DID
+
+```jsx
+cacaoCID = ipld.put(CACAO(SIWx-message, signature))
+did = 'did:3:' + multihash(cacaoCID)
+```
+
+#### Read
+
+To read you must load a deterministic Ceramic stream to find any updates made to the 3ID.
+
+1. Load the revocation registry Ceramic stream
```

Comment:
It's not only used to revoke verification methods. Any CACAO can be revoked using this registry. The main reason you would register a CACAO but not revoke it is to be able to have chained revocations not be recursively revoked. So in my mind this is still mainly a revocation registry.

Reply:
but it can also be used not for revocation at all.  It is also a registry of the **valid** verification methods.  For 3ids that have never done a revocation this stream will contain no revocations at all, only additions of new valid VMs.  Or am I misunderstanding?

Reply:
I think I don't understand what this means:
> The main reason you would register a CACAO but not revoke it is to be able to have chained revocations not be recursively revoked

don't you add an entry for every new wallet/key you add to the 3id?